src/irmin/store_intf.ml

```diff
@@ -429,8 +429,14 @@ module type S = sig
         operations as the generating computation, in the exact same order.*in
         some order*. *)
 
-    val produce_stream : (stream_proof, 'a) producer
-    (** [produce_stream] is the producer of stream proofs. *)
+    val produce_stream :
+      ?with_extenders:bool ->
+      repo ->
+      kinded_hash ->
+      (tree -> (tree * 'result) Lwt.t) ->
+      (stream_proof * 'result) Lwt.t
+    (** Same as {!produce_proof} but for stream proofs. Note that if
+        [with_extenders] is set, the read order is no longer guaranteed. *)
```

Comment:
What does it mean that the reading order is not preserved? Do you have an example? That doesn't seem like a good fit for stream proofs...

Reply:
Say you have an inode that contains 2 buckets with extenders. The information on which of these 2 buckets was loaded first is lost.

Reply:
When we read the root inode we don't know yet which paths can be compressed into extenders - we only know this after we read all needed values from that inode - so the read order is not preserved when compressing the path into extenders. For example in the stream
```
{"Inode":{"length":3,"proofs":[[[0],".."]]}},
{"Inode":{"length":3,"proofs":[[[0],"..."]]}},
{"Inode":{"length":3,"proofs":[[[0],"..."]]}} 
```
we can compress to [0;0;0] - but I think we don't have any guarantee that there are no other reads interleaved with the ones for path [0;0;0].

Reply:
Consider this inode tree:
```
------------------
|  ptr   | ptr   |   Tree `p`
------------------
   |            \
   |             \__________________------------------
   |                                |   ptr  |  None |  Tree `b0`
   |                                ------------------ 
------------------                       |
|  None   | ptr  |   Tree `a0`           |
------------------                     Values `b1`
             |                                
             |                                
             |
             |
         Values `a1`
```

Both these "pre"-streams:
- `p, a0, a1, b0, b1`
- `p, b0, a0, a1, b1` (This situation could happen if the `f` function attemps a read to an unbound `path`)

are converted to this stream:
- `p, a1, b1`

Reply:
I think what we care about is the order in which we read Inode values, e.g. intermediate nodes are not super important as the user can't control them with a high-level call (e.g. users will only do `Tree.find t k`, `Tree.add t k v`, etc.

So as long as we are not exchanging values in the trace I think we are fine. WDYT?

Reply:
I don't think it matters a lot, but it may depend on the way it is used by ORUs. 

Providing an `?with_extenders` parameter allows us to cover both situations

Reply:
Hum interesting...

So:

- `p`, `b0`, `a0`, `a1`, `b1` (This situation could happen if the f function attemps a read to an unbound path)
are converted to this stream: `p`, `a1`, `b1`

Does this mean that:
```
let f t = 
  let* x = Tree.find t ["11"] in
  let* y = Tree.find t ["01"] in
  let+ z = Tree.find t ["10"] in
  assert (x = None);
  assert (y = Some "a1");
  assert (z = Some "b1")
```
will fail to verify?

Reply:
I think it will verify fine. The problem is that the following code will produce the exact same proof:

```
let f t = 
  let* y = Tree.find t ["01"] in
  let+ z = Tree.find t ["10"] in
  assert (y = Some "a1");
  assert (z = Some "b1")
```


PS: I'm not totally sure

Reply:
In that case it's fine!